log/loki.go

```diff
@@ -0,0 +1,419 @@
+			switch value {
+			case "debug":
+				h.levels = append(h.levels, logrus.DebugLevel)
+				fallthrough
+			case "info":
+				h.levels = append(h.levels, logrus.InfoLevel)
+				fallthrough
+			case "warning":
+				h.levels = append(h.levels, logrus.WarnLevel)
+				fallthrough
+			case "error":
+				h.levels = append(h.levels, logrus.ErrorLevel)
+			default:
+				return nil, fmt.Errorf("unknown log level %s", value)
+			}
+		default:
+			if strings.HasPrefix(key, "label.") {
+				labelKey := strings.TrimPrefix(key, "label.")
+				h.labels = append(h.labels, [2]string{labelKey, value})
+				continue
+			}
+
+			return nil, fmt.Errorf("unknown loki config key %s", key)
+		}
+	}
+
+	h.start()
+	return h, nil
+}
+
+func (h *lokiHook) start() {
+	h.ch = make(chan *logrus.Entry, 1000)
+	go h.loop()
+}
+
+// fill one of two equally sized slices with entries and then push it while filling the other one
+// TODO clean old entries after push?
+func (h *lokiHook) loop() {
+	var (
+		msgs       = make([]tmpMsg, h.limit)
+		msgsToPush = make([]tmpMsg, h.limit)
+		dropped    int
+		count      int
+		ticker     = time.NewTicker(h.pushPeriod)
+		pushCh     = make(chan chan struct{})
+	)
+
+	defer close(pushCh)
+
+	go func() {
+		oldLogs := make([]tmpMsg, 0, h.limit*2)
+		for ch := range pushCh {
+			msgsToPush, msgs = msgs, msgsToPush
+			oldCount, oldDropped := count, dropped
+			count, dropped = 0, 0
+			close(ch) // signal that more buffering can continue
+
+			copy(oldLogs[len(oldLogs):len(oldLogs)+oldCount], msgsToPush[:oldCount])
+			oldLogs = oldLogs[:len(oldLogs)+oldCount]
+			n, err := h.push(oldLogs, oldDropped)
+			_ = err // TODO print it on terminal ?!?
+			if n > len(oldLogs) {
+				oldLogs = oldLogs[:0]
+				continue
+			}
+			oldLogs = oldLogs[:copy(oldLogs, oldLogs[n:])]
+		}
+	}()
+
+	for {
+		select {
+		case entry, ok := <-h.ch:
+			if !ok {
+				// TODO push one last time ?
+				return
+			}
+
+			if count == h.limit {
+				dropped++
+				continue
+			}
+
+			// Arguably we can directly generate the final marshalled version of the labels right here
+			// through sorting the entry.Data, removing additionalparams from it and then dumping it
+			// as the final marshal and appending level and h.labels after it.
+			// If we reuse some kind of big enough `[]byte` buffer we can also possibly skip on some
+			// of allocation. Combined with the cutoff part and directly pushing in the final data
+			// type this can be really a lot faster and to use a lot less memory
+			labels := make(map[string]string, len(entry.Data)+1)
+			for k, v := range entry.Data {
+				// TODO filter the additional ones
+				labels[k] = fmt.Sprint(v) // TODO optimize ?
+			}
+			for _, params := range h.labels {
+				labels[params[0]] = params[1]
+			}
+			labels["level"] = entry.Level.String()
+			// TODO have the cutoff here ?
+			// if we cutoff here we can cut somewhat on the backbuffers and optimize the inserting
+			// in/creating of the final Streams that we push
```

Comment:
:+1: The data passed to `push()` should preferably be ready for pushing, and any pre-processing should happen outside of `push()`.

Same for the sorting, maybe. Could we sort messages as they're added to `msgs`, using something like a priority queue or a self-balancing tree? Or would that perform worse in this case and not be worth the complexity?

Some benchmarks would be nice, though I suspect the bulk of the time is spent on network I/O.

Reply:
I did some basic timing of stuff and:
```
sorting=74.345524ms, adding=72.565886ms marshalling=145.5386ms sending=1.740313494s count=184143 final_size=6623126
```
So most of the time in this case is spend in loki ... ingesting? I probably should try to configure loki to be more performant